rust/src/model.rs

```diff
@@ -0,0 +1,160 @@
+use std::convert::TryFrom;
+
+use num::{bigint::BigInt, clamp, rational::Ratio, traits::float::FloatCore};
+use rand::SeedableRng;
+use rand_chacha::ChaCha20Rng;
+
+use crate::{
+    mask::{
+        config::{DataType, MaskConfig},
+        seed::MaskSeed,
+        MaskIntegers,
+        MaskedModel,
+    },
+    utils::generate_integer,
+    PetError,
+};
+
+#[derive(Clone, Debug, PartialEq)]
+/// A model with parameters represented as a vector of numbers.
+pub enum Model {
```

Comment:
Why using an enum here? My first impression is that generics would fit nicely:

```rust
struct Model<T>(Vec<T>)
```

The only method that must be implemented manually for each type afaict is `xxx_as_ratio` so we could have:

```
trait AsRatios {
    fn as_ratios(&self) -> Option<Vec<Ratio<BigInt>>>;
}

impl AsRatios for Model<f64> { ... }
impl AsRatios for Model<f32> { ... }
impl AsRatios for Model<i32> { ... }
// etc.
```

Then:

```impl<T> Model<T>
where
    Model<T>: AsRatios,
{
    fn mask(&self, ...,) {
	    // less repetition now
    }
}
```

I may be missing something though.

Reply:
i started with generics here, but failed. the issue is that the methods from the `num` crate impose some specific trait bounds on their method arguments. for example methods on floats are bounded by `Float`/`FloatCore` whereas methods on integers are bounded by `Integer`/`PrimInt` and those are disjunct. any generic type `struct Model<T>(Vec<T>)` must impose either one of the trait bounds on `T` at the same time. my workaround was to implement it as an enum without generics.

Reply:
Ah then maybe what we need is different impls for floats and ints. Basically `Model<T: Float>` and `Model<T: Integer>` ? I'd need to take a closer look though.

Reply:
i refactored it using traits now. i was basically blocked by having an unnecessary extra layer of generics in my mental model, like `primitive_type: FloatCore/PrimInt: Model<N>`. by resolving that with the new traits and removing most of the occurrences of `FloatCore`/`PrimInt` this works now. thanks for pointing me in the right direction!